### PR TITLE
Remove trivial statements before evaluating groundedness

### PR DESCRIPTION
**Pull request by sfc-gh-jreini**
# Description

As many users have noted, not all statements from an LLM _should_ be evaluated for groundedness, such as stylized messages and trivial statements.

This change will remove such statements before evaluating for groundedness.

This prevents the score from being diluted by these statements, and removes false hallucination detection for statements not expected to be grounded (because they're trivial).

## Example:

"Hi. I'm here to help. The university of Washington is a public research university. UW's connections to major corporations in Seattle contribute to its reputation as a hub for innovation and technology"

The function will split the statement into its component sentences:

1. "Hi."
2. "I'm here to help."
3. "The university of Washington is a public research university."
4. "UW's connections to major corporations in Seattle contribute to its reputation as a hub for innovation and technology"

Next, trivial statements are removed, leaving only:

3. "The university of Washington is a public research university."
4. "UW's connections to major corporations in Seattle contribute to its reputation as a hub for innovation and technology"

The LLM will then process the statement, to assess the groundedness of the statement.

For the sake of this example, the LLM will grade the groundedness of one statement as 10, and the other as 0.

Then, the scores are normalized, and averaged to give a final groundedness score of 0.5.

## Before
Screenshot 2024-08-20 at 5 58 16 PM

## After

Note: both LLM responses start with "Hi. I'm happy to help with your question.." But with this change, these first two sentences are excluded from the evaluation.
Screenshot 2024-08-20 at 6 52 03 PM

## Other details good to know for developers

Please include any other details of this change useful for _TruLens_ developers.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] New Tests
- [ ] This change includes re-generated golden test results
- [ ] This change requires a documentation update

### CODE REVIEW

File: src/feedback/trulens/feedback/llm_provider.py [hunk 1]
```diff
@@ -1314,14 +1313,43 @@ def stereotypes_with_cot_reasons(
...
         self, source: str, statement: str
     ) -> Tuple[float, dict]:
         """A measure to track if the source material supports each sentence in
         the statement using an LLM provider.
 
-        The LLM will process the entire statement at once, using chain of
-        thought methodology to emit the reasons.
+        The statement will first be split by a tokenizer into its compoenent sentences.
+
+        Then, trivial statements are eliminated so as to not delute the evaluation.
```
Comment (sfc-gh-pdharmana) [inline src/feedback/trulens/feedback/llm_provider.py:1350]: dilute?

File: src/feedback/trulens/feedback/v2/feedback.py [hunk 2]
```diff
@@ -254,6 +254,17 @@ class Abstention(Semantics, WithPrompt):
     )
 
 
+class Trivial(Semantics, WithPrompt):
+    system_prompt: ClassVar[str] = cleandoc(
+        """Consider the following list of statements. Remove any statements that are trivial, irrelevant, or unimportant. Respond with a list of the remaining statements."""
+    )
```
Comment (sfc-gh-pdharmana) [inline src/feedback/trulens/feedback/v2/feedback.py:260]: We did not clearly define in this prompt, what a trivial statement is...

File: src/feedback/trulens/feedback/llm_provider.py [hunk 3]
```diff
@@ -1314,14 +1313,43 @@ def stereotypes_with_cot_reasons(
 
         return self.generate_score_and_reasons(system_prompt, user_prompt)
 
+    def _remove_trivial_statements(self, statements: List[str]) -> List[str]:
```
Comment (sfc-gh-pdharmana) [inline src/feedback/trulens/feedback/llm_provider.py:1316]: Would be good to add any benchmarking results post this change. @sfc-gh-dhuang pls help with that.

Comment (sfc-gh-dhuang) [reply src/feedback/trulens/feedback/llm_provider.py:1316]: Yes can do. Time to start a backlog of things that we wanna meta-evaluate later after the PrPr :D

File: src/feedback/trulens/feedback/v2/feedback.py (hunk 2 above)
Comment (sfc-gh-dhuang) [reply src/feedback/trulens/feedback/v2/feedback.py:260]: suggestion 
```
"Identify and remove sentences that are stylistic, contain trivial pleasantries, or lack substantive information relevant to the main content."
```

Credits to ChatGPT

**Comment by sfc-gh-dhuang** [comment]
Add one suggestion for more detailed / focused prompt. Other than that LGTM and thanks for taking care of this Josh! 🚀

**Review by sfc-gh-dhuang — COMMENTED** [review]
Add one suggestion for more detailed / focused prompt. Other than that LGTM and thanks for taking care of this @sfc-gh-jreini ! 🚀

**Review by sfc-gh-dhuang — APPROVED** [review]

**Pull request merged**